### PR TITLE
Fix polling in WasiPipe

### PR DESCRIPTION
**Pull request by fschutt**
> @john-sharratt: the polling functionality you have implemented won't work as its not 
> actually polling the receiver - you need to implement it like this: 
> https://github.com/john-sharratt/wasmer/blob/fe14f086808431d6f64b540421798ac779cdce0a/lib/wasi/src/state/pipe.rs#L223
>
> Note: I've converted wasmer-os over to wasmer and now using WasiPipe hence its important this is implemented correctly. Especially as bash uses Polling on stdin which is connected via this pipe.

### CODE REVIEW

**Comment by fschutt** [comment]
TODO: when this is done and bash / dash is working, we can test it with `echo blah | cat`

File: lib/wasi/src/state/pipe.rs [hunk 1]
```diff
@@ -235,18 +235,27 @@ impl WasiPipe {
...
-                    let reader = buf.as_ref();
-                    let read = read_bytes(reader, memory, iov).map(|_| buf_len as usize)?;
-                    buf.advance(read);
-                    return Ok(read);
+            if let Ok(mut read_buffer) = self.read_buffer.lock() {
+                if let Some(buf) = read_buffer.as_mut() {
+                    let buf_len = buf.len();
+                    if buf_len > 0 {
+                        let reader = buf.as_ref();
+                        let read = read_bytes(reader, memory, iov).map(|_| buf_len as usize)?;
```
Comment (theduke) [inline lib/wasi/src/state/pipe.rs:243]: Seems like you are throwing away the actual amount of bytes read, which can be lower than the buffer size.

File: lib/wasi/src/state/pipe.rs [hunk 2]
```diff
@@ -305,19 +316,21 @@ impl Seek for WasiPipe {
...
-                        let read = reader.read(buf).map(|_| buf_len as usize)?;
-                        inner_buf.advance(read);
-                        return Ok(read);
+            if let Ok(mut read_buffer) = self.read_buffer.lock() {
+                if let Some(inner_buf) = read_buffer.as_mut() {
+                    let buf_len = inner_buf.len();
+                    if buf_len > 0 {
+                        if inner_buf.len() > buf.len() {
+                            let mut reader = inner_buf.as_ref();
+                            let read = reader.read_exact(buf).map(|_| buf.len())?;
```
Comment (theduke) [inline lib/wasi/src/state/pipe.rs:325]: Same thing as above.

File: lib/wasi/src/state/pipe.rs [hunk 3]
```diff
@@ -305,19 +316,21 @@ impl Seek for WasiPipe {
...
+                            inner_buf.advance(read);
+                            return Ok(read);
+                        } else {
+                            let mut reader = inner_buf.as_ref();
+                            let read = reader.read(buf).map(|_| buf_len as usize)?;
```
Comment (theduke) [inline lib/wasi/src/state/pipe.rs:330]: As above.

**Review by theduke — COMMENTED** [review]
Not familiar enough with this code to provide a comprehensive review.

Did notice one problematic pattern, though.

**Comment by theduke** [comment]
This will be subsumed by #3422 .

**Pull request closed without merging**